metadetect/metadetect.py

```diff
@@ -115,57 +201,261 @@ def go(self):
+                all_is_shear_band.append(False)
+                band_mbobs_list = []
+                for mbobs in nonshear_mbobs_list:
+                    _mbobs = ngmix.MultiBandObsList()
+                    _mbobs.meta = copy.deepcopy(mbobs.meta)
+                    _mbobs.append(mbobs[band])
+                    band_mbobs_list.append(_mbobs)
+                band_res.append(self._fitter.go(band_mbobs_list))
+
+        # combine the data by inverse variance weighted averages
+        tot_nband = self.nband + self.nonshear_nband
+        n = Namer(front=self['model'])
+        dt = [
+            ("flags", 'i4'),
+            (n("flags"), 'i4'),
+            (n("s2n"), "f8"),
+            (n("T"), "f8"),
+            (n("T_err"), "f8"),
+            (n("g"), "f8", 2),
+            (n("g_cov"), "f8", (2, 2)),
+            ('psf_g', 'f8', 2),
+            ('psf_T', 'f8'),
+            (n("T_ratio"), "f8"),
+        ]
+        if tot_nband > 1:
+            dt += [
+                (n("band_flux"), "f8", tot_nband),
+                (n("band_flux_err"), "f8", tot_nband),
+            ]
+        else:
+            dt += [
+                (n("band_flux"), "f8"),
+                (n("band_flux_err"), "f8"),
+            ]
+
+        res = np.zeros(len(mbobs_list), dtype=dt)
+        for ind in range(len(mbobs_list)):
+            # extract the wgts and band results
+            wgts = []
+            all_bres = []
+            for i, obslist in enumerate(mbobs_list[ind]):
+                msk = obslist[0].weight > 0
+                if not np.any(msk):
+                    raise RuntimeError("No non-zero weights for a band in metadetect!")
+                wgts.append(np.median(obslist[0].weight[msk]))
+                all_bres.append(band_res[i][ind:ind+1])
+
+            if nonshear_mbobs_list is not None:
+                for i in range(self.nband, tot_nband):
+                    all_bres.append(band_res[i][ind:ind+1])
+                    wgts.append(1)
+
+            wgts = np.array(wgts)
+            wgts[0:self.nband] = wgts[0:self.nband] / np.sum(wgts[0:self.nband])
+
+            shear_flags = 0
+            for i in range(self.nband):
+                shear_flags |= all_bres[i]['flags'][0]
+
+            # compute the averages
+            band_flux = []
+            band_flux_err = []
+            flux = 0.0
+            flux_var = 0.0
+            for wgt, bres, is_shear_band in zip(wgts, all_bres, all_is_shear_band):
+                res['flags'][ind] |= bres['flags'][0]
+                res[n('flags')][ind] |= bres['flags'][0]
+
+                if is_shear_band and shear_flags == 0:
+                    flux += (wgt * bres[n('flux')][0])
+                    flux_var += (wgt * bres[n('flux_err')][0])**2
+
+                    res[n('T')][ind] += (wgt * bres[n('T')][0])
+                    res[n('T_err')][ind] += (wgt * bres[n('T_err')][0])**2
+
+                    res[n('g')][ind] += (wgt * bres[n('T')][0] * bres[n('g')][0])
```

Comment:
I agree with this averaging of moments rather than `g` values.

However, we may not want to attempt to form `g` at all on the individual bands because that can fail if a given band is noisy, but the combined can succeed.  Or at least ignore such errors and only check the final `T`

Reply:
e.g. if `T` is noisy in one band, you might get `T <= 0` and you cannot form `M1/T`; I don't know if you are checking for such an error on each band and then failing the object